### PR TITLE
Gamerules minimum viable product

### PR DESCRIPTION
**Pull request by Amazinite**
**Feature:** This PR implements the feature request detailed and discussed in issue #5556.

## Feature Details
This PR is intended to be a minimum viable product for gamerules. That is, this implements only the bare minimum necessary to use as a foundation for future improvements. With this minimum viable product, the "gamerules" node has the current behaviors:
* A `gamerules` node changes the default gamerule values for all save files.
* Gamerule cannot be modified in any way once the game has started.
* The following two gamerules are implemented:
  * `"universal ramscoop" (true | false)`: Changes whether the universal ramscoop is active. Default true.
  * `"person spawnrate" (N >= 1)`: Changes the spawnrate of person ships. Default 36,000 frames (10 minutes).
  * `"no person spawn weight" (N >= 0)`: Changes the weight for no person ship to spawn when a spawn attempt is made. Default 1000 frames.

This is mostly resurrected from #5978, although pared down to only three gamerules, and avoiding the use of a Dictionary as quyykk suggested.

## UI Screenshots
N/A

## Usage Examples
The following are the default values for the two gamerules currently implemented.
```
gamerules
	"universal ramscoop" true
	"person spawnrate" 36000
```
Gamerules will be either integer rules, decimal rules, or boolean rules. Boolean rules, such as "universal ramscoop" accept values of "true" and "false." Integer rules such as "person spawnrate" accept whole numbers. Decimal rules, of which there are currently none, would accept, well, decimals. Each rule could be given its own additional constraints. For example, "person spawnrate" must be >= 1.

The `gamerules` node does not overwrite each time it is read. What that means is that if you only wanted to change a single gamerule, all you would need is to add a file to a plugin with the following text:
```
gamerules
	"universal ramscoop" false
```
In the above case, this would change the "universal ramscoop" rule to false and make no other changes to gamerules.

## Testing Done
Tested the universal ramscoop rule while true and false. While true, received fuel even without a ramscoop. While false, did not.
Tested the person spawnrate rule with a much lower value than the default, which resulted in many more person ships appearing.

### Automated Tests Added
Created unit tests for the new IsBool and BoolValue methods of DataNode.

## Performance Impact
N/A

## Future Work
Since this PR is only the bare minimum of what a gamerules mechanic could provide, here are my notes on where this should go in the future. These notes will not include new gamerules that we could add, but instead new ways in which the system itself could be expanded:
* Allow gamerules to be defined on a per-save file basis. This would entail having the gamerules written to the save file, and once they are read during loading, those values would be what gets used.
* Allow starts to define which gamerules get used. This would entail having a `gamerules` node under `start`. Any gamerules defined in a start should supersede those rules defined by any `gamerules` root nodes, i.e. the default gamerules.
* Either as an alternative or in addition to starts defining gamerules, allow for the definition of "gamerule sets" that can be chosen independently of a start. This would allow for something like first choosing your starting location, then choosing the difficulty you want to play at.
* A step beyond just having starts define gamerules, create some UI that can be accessed during choosing your start that allows for the tweaking of specific gamerules. This would allow for further fine-tuning your experience beyond just what starts and gamerule sets allow, although this should definitely be marked as an "advanced setting" type of thing, and probably wouldn't allow for the tweaking of every single gamerule.
* Create prefixed conditions that allow for the querying of the states of gamerules. This could allow missions and other parts of the game to react to which gamerules are in play. For example, if combined with conditional outfit descriptions, the ramscoop description could mention the existence of the universal ramscoop, but only if you have that gamerule set to true.

Some random implementation ideas:
* One problem that needs solved with the above is how one might stack multiple sets of gamerules on top of one another. Perhaps a Rule class could exist, where a Rule has some enum that denotes what type its value is, be that bool, int, or double. It could then contain another enum noting the state of the Rule, be that undefined, default, or modified. If combining two gamerule sets, modified values would replace default and other modified values, and default values would replace undefined values. This would simplify having to have some labyrinthine if/else structure that checks each individual gamerule for if it should be replaced or not when combining sets. Then a GameruleSet could be a subset of Rules, with GameruleSets being what are defined by starts or difficulty levels, while the Gamerules class is every possible Rule, either default or modified (i.e. no undefined rules), and is what gets used by the game.

### CODE REVIEW

**Comment by Zitchas** [comment]
Nice to see gamerules arriving. Probably a good pair to start with.

For lore reasons I'd prefer the universal ramscoop to default to false, but either way, I'm happy to have it in here.

**Comment by Terin** [comment]
Great to see this at last. Definitely would prefer starting from here and slowly adding in other constants over time, on top of that additional functionality.

File: EndlessSky.xcodeproj/project.pbxproj
Comment (RisingLeaf) [inline EndlessSky.xcodeproj/project.pbxproj:30]: this change seems unrelated?

File: source/Gamerules.cpp
Comment (RisingLeaf) [inline source/Gamerules.cpp:26]: Actually those could become useful on all tokens, so why dont we integrate them into the datanode class?

**Review by RisingLeaf — APPROVED** [review]
I like it, seems pretty straight forward, no complains

File: EndlessSky.xcodeproj/project.pbxproj
Comment (warp-core) [reply EndlessSky.xcodeproj/project.pbxproj:30]: The xcode patch I assume Derpy has used sometimes does that, I think.

Comment (RisingLeaf) [reply EndlessSky.xcodeproj/project.pbxproj:30]: ok

File: data/gamerules.txt
Comment (Hurleveur) [inline data/gamerules.txt:23]: should it be `spawn rate`?

File: source/Gamerules.h
Comment (Hurleveur) [inline source/Gamerules.h:37]:
```suggestion
	int personSpawnrate = 36000;
	bool universalRamscoop = true;
```
So they're in the same order as for the methods

File: source/Gamerules.h
Comment (Hurleveur) [inline source/Gamerules.h:21]:
```suggestion


```

**Review by Hurleveur — CHANGES_REQUESTED** [review]
I agree it's best to have it this simple to begin with.
Also frankly idk if we need a test for it, though we could make a ship without ramscoop idle a bit in a system with a powerful star and 0.99 units of fuel, trying to jump all the time until it tries to land. But really there's no need.

Looking forward to expand upon this!

File: data/gamerules.txt
Comment (1010todd) [inline data/gamerules.txt:22]:
```suggestion
	# A person ship spawns on average every this many frames. The default is once every ten minutes (36000 frames).
```
Add default frame to the comment incase someone accidentally replaced it and forgot the original value. Hopefully will save a few questions online.

File: EndlessSky.xcodeproj/project.pbxproj
Comment (quyykk) [reply EndlessSky.xcodeproj/project.pbxproj:30]: @RisingLeaf check this one out https://github.com/endless-sky/endless-sky/pull/7612/files#diff-f99692e65fcaf0033fc339558d87b2a512a27f2396e27bb6932887744392d967

File: source/Ship.cpp
Comment (quyykk) [inline source/Ship.cpp:2358]: It's intentional that you're using a different formula for the universal ramscoop here right?

File: source/Gamerules.h
Comment (Zitchas) [inline source/Gamerules.h:36]:
```suggestion
	bool universalRamscoop = false;
```

Universal ramscoop is a lore issue, not just a convenience/safety factor. Giving our players the option to have it enabled is good; but the default state should be without it.

**Review by Zitchas — CHANGES_REQUESTED** [review]
Awesome to see this moving forward. Definitely hoping that this makes it into 0.9.17.

Thanks for picking this up again!

File: source/Gamerules.h
Comment (Hurleveur) [reply source/Gamerules.h:36]: This is something that will be trivial to change by anyone of the community when we do add some mechanics to compensate. The last spaceport save is something in the right direction, but not sufficient.
Also since an issue was open instantly to change that, this obviously isn't a change everyone would agree to.

The best example of "lore" change that ruined the game experience would be the removal of portraits.

File: source/Ship.cpp
Comment (Amazinite) [reply source/Ship.cpp:2358]: No, just accidentally missed a 0.

File: source/Gamerules.h
Comment (Zitchas) [reply source/Gamerules.h:36]: We have *four* backup saves to prevent getting locked out, and better safety mechanisms have been discussed and are hopefully being worked on that make for *better* gameplay.

I'll note that in EV/O/N running out of fuel after having landed on a *single* uninhabited planet was a hard lock that led to player's having to either edit their save (not very easy to do) or start a new pilot, and yet it remained a staple of the franchise through two additional sequels. Given the volumes of people who successfully played the trilogy and loved it, I really don't think our new players need this hand holding. Especially since pretty much everywhere a new player can reach in the first few hours (baring considerable luck) is going to have *plenty* of traffic to help get them unstuck in an immersive, in-character, way. New players are no less capable of us, and depriving them of the opportunity to overcome this is worse than it'd be to remove pirates from the early game. Worse, because fuel is a challenge that a new player *can* overcome, whereas pirates are something they'll usually have to just run from for quite a while. Having meaningful challenges that they can overcome *is* an important part of gameplay. Yes, it is an easy one, as befits something that should mainly be a challenge for the very early game. (there are, after all, *lots* of ways to mitigate or even eliminate this risk, some of which are available from day 1.

File: data/gamerules.txt
Comment (Amazinite) [reply data/gamerules.txt:23]: Probably, but it's one of those pairs of words to me that just seems like eventually it'll become one word, like how web site became website. There's a few other such pairs of words that I've been badgered about sometimes pressing together, like eachother or everytime. (Like, come on, at least with everytime, everyone and everywhere are already accepted single words.) Gamerule should also technically be "game rule," although that one being one word is literally just because the command in Minecraft is "gamerule."

File: source/Gamerules.h
Comment (Hurleveur) [reply source/Gamerules.h:36]: To be clear I'm not saying no never, I'm saying first have more spawnratea when low on fuel or anything else really to compensate

File: EndlessSky.xcodeproj/project.pbxproj
Comment (RisingLeaf) [reply EndlessSky.xcodeproj/project.pbxproj:30]: Uh, you know, I am working with xcode amd it does that automatical, yeah I get it, i should not have complained

File: source/Gamerules.h
Comment (Zitchas) [reply source/Gamerules.h:36]: The community doesn't have a *concensus* on the change, agreed. However, I think it's a change that we should keep and people should actually try out, rather than just blindly revert.

I'm quite serious when I say that I feel this makes the game *better.* As in all-round better gameplay, better experience, better story, better challenge, better respect for the player.

There's a reason why not only it was the case in EV, it was *kept* through two sequels, even though both EVO and EVN had *far* more uninhabited systems that one could get stuck in. A lot more long stretches that required ships to actually make an *effort* to have above average fuel capacity, too. They had the programming ability to change it if people running out of fuel had been a problem.

Comment (Amazinite) [reply source/Gamerules.h:36]: The point of gamerules is to customize one's own experience. If *you* think that this gamerule is better set to false then *you* should customize your experience to do so. But this behavior has been true since v0.9.0, so that, plus the non-insignificant number of people who commented after #7730 was merged, suggests to me that with the introduction of the gamerule, its state should default to true.

Remember too that this PR is only a foundation. While at the moment it won't be easy for everyone to change this gamerule, it being a gamerule at least makes it several times more customizable than before, and it will become easier for people to mess with in time as gamerules start to be able to be customized in more places.

I also think it's important to note, because you do this a lot, that while Endless Sky is inspired by Escape Velocity, it is not itself Escape Velocity. Sure, we can look to EV for how it did things, but I don't think we should say "Because EV did this, we should do this." Now you might not be saying that exactly, but you do often use the fact that EV did things a certain way as part of an argument for why we should do things a certain way. We are a different game with different systems. Some similar, some not so similar, but at the end of the day we are a different game. Just because EV did something doesn't mean it's better and therefore should be done here. Any changes made to ES need to be justified *within* ES.

Comment (Terin) [reply source/Gamerules.h:36]: Also, getting trapped on Vellos sucked. Getting trapped on Kontik sucked. Even if EV Nova had multiple previous saves, those are unnecessary gotchas for new players. Having the emergency ramscoop doesn't mean players won't use ramscoops later on, especially with out deep in finding the Ember Wastes is for a normal playthrough.

I'm kinda shocked to find out that removing the ramscoop was strong-armed like this. Default it on. I still had a wonderful time exploring the Ember Wastes the first time. Having a very minor safety net did not impact my enjoyment of it at all.

Comment (Zitchas) [reply source/Gamerules.h:36]: I reference EV/O/N because it is over two decades of combined gameplay that we can learn from. We're already significantly better than it in many ways, but fuel management isn't one of them. Creating methods to prevent lock-out? Great! Awesome! I'm all in favor of that. Giving all ships infinite fuel is a *terrible* way to do that.

And we've already fixed that so players *can't* get stuck out there. Vellos and Kontik player-death literally can't happen in ES now. There's no need to hand-hold them the entire way.

And once again, I'll just point out: I'm a content creator writing lore for the story. I have to take into account the typical environment people play in while writing that story. Fuel scarcity is, actually, going to be an upcoming story arc. Or at least, that's what I have roughed out. Universal ramscoop makes a mockery of that. I'm not going to create a situation where the player can go "Hey, no problem, I'll just take off, make you 20-30 jumps of fuel, and here you go." Hey, if I can do that, why can't you? Hey, I must be special or something!"

File: source/Gamerules.h
Comment (Amazinite) [reply source/Gamerules.h:37]: Swapping the methods instead of the variables, because the variables are in the order I have them in the text file.

File: source/Gamerules.h
Comment (Zitchas) [reply source/Gamerules.h:37]: Beyond that, whatever happened to "Let's just let it sit and see what people think after experiencing it?" This is, after all, the continuous build. It's not even scheduled to be pushed out to an *unstable* release untill... January, maybe? Possibly February? There's a *lot* of time to have a month or two to let people actually try it out, implement various extra safety things, etc.

And before you say "they can try it out as a gamerule." I'll just state that applies just as much to trying out PRs, and we all know that doesn't happen. As Amazinite pointed out on Discord (roughly paraphrasing): "No matter how long a PR sits, the most likely time for negative reactions is immediately after it is merged, because *that* is when all the eyes actually look at how things changed." So let's *not* just leap to roll back the change into an invisible setting that the majority of people will never know is there and never change, and see how people actually handle it. See if an of the people on the cutting edge of continuous *actually* get stuck. I know I haven't so far (despite running out of fuel in uninhabited systems about a dozen times now, not a single reload because of it.).

File: source/DataNode.cpp
Comment (quyykk) [inline source/DataNode.cpp:257]: Should we also accept the use of `1` and `0` instead of `true` and `false`?

**Review by quyykk — COMMENTED** [review]
just this one comment :smile:

File: source/DataNode.cpp
Comment (Amazinite) [reply source/DataNode.cpp:257]: I could see that being done. In the OP, I mentioned this future enhancement:
> Create prefixed conditions that allow for the querying of the states of gamerules. 

Querying a boolean rule would have to return 1 or 0, since conditions are ints. So in that sense, it might make sense to support 1 and 0. But in another sense, I only expect boolean gamerules to be defined with "true" or "false" as their values, not "1" or "0," so is there any reason aside from symmetry with conditions that we'd do support that?

File: source/Ship.cpp
Comment (Amazinite) [inline source/Ship.cpp:2355]:
```suggestion
			// If the universal ramscoop gamerule is true, then even if a ship has no ramscoop,
```

File: source/DataNode.cpp
Comment (quyykk) [reply source/DataNode.cpp:257]: I was thinking about changing existing code to use these booleans (for example, preferences could use that instead of 0 and 1 like they do currently for some). If we were to do that we'd of course need to parse 0/1 as well to maintain backwards compatibility.

File: source/DataNode.cpp
Comment (Amazinite) [inline source/DataNode.cpp:245]:
```suggestion
// Check if the token at the given index is a boolean, i.e. "true"/"1" or "false"/"0"
```

File: source/DataNode.cpp
Comment (Amazinite) [reply source/DataNode.cpp]: Added 1 and 0 as valid forms of true and false.

File: source/Gamerules.h
Comment (Quantumshark) [reply source/Gamerules.h:37]: From the lore perspective, ramscoop outfits aren't exactly scarce - a 10 ton outfit available in human space from the start of the game for 60,000 credits might be outside the scope for a new player who is working off their loan (which is an important demographic, possibly the most important, to consider with regards to universal ramscoop) or a player still flying a smaller ship without 10 tons to spare, but it's not something that governments, factions or even well-off individual citizens would struggle to get their hands on.
And that's before considering that quite a few ships have built-in ramscoops anyway. So I don't understand how universal ramscoop (which is still reduced to 0 in any system where the effect of a ramscoop outfit would be reduced to 0 - it doesn't ignore changes to solar wind) could break a planned arc around fuel scarcity without ramscoop outfits or the built-in ramscoop of, for example, every korath automata warship also breaking it.

**Comment by Zitchas** [comment]
No, our descriptions should be broadly applicable, not specific to a particular case. That explanation could go in a tutorial pop-up or something that only triggers when the gamerule is set accordingly.

File: source/Gamerules.h
Comment (1010todd) [reply source/Gamerules.h:37]: I feel like we want to just get gamerule code in first so other PRs could utilize it. The universal ramscoop is clearly too controversial atm and it will just unnecessarily block this useful code feature. Probably best to just keep it to what's already in the game and discuss about removing or keeping it somewhere else.

File: source/DataNode.cpp
Comment (Hurleveur) [inline source/DataNode.cpp:246]: This gets me thinking; dont we use never as a bool to say false?

File: source/Gamerules.h
Comment (Hurleveur) [reply source/Gamerules.h:37]: Is it controversial though? I feel like only Zitchas wants it off
And since realism is not a valid reason it should be on.
 If you want to remove it permanently go open an issue about it like smbd did to bring it back. I dont expect many will be for it, it has already worsened my game experience for sure.
And it seems like what Zitchas wants is just to disable it for the ember wastes not for the whole game. I could easily code that later on.

**Review by Hurleveur — APPROVED** [review]
Just a small comment on booleans but this looks good to me anyway since those are enot the focus of this PR.

File: source/Gamerules.h
Comment (1010todd) [reply source/Gamerules.h:37]: It sparked entire poll to be made about merging process as well as enough talk here for Github to collapse some of the conversation. Personally I don't care too much about it except now it's blocking a generally useful PR. I just want to see this PR in first so other gamerule stuffs can follow and people who want to fight about the ramscoop thing can go do it somewhere else. 
Overriding gamerule per system sounds interesting though, but also sounds a little odd when its a universal rule and somehow these systems just say no but that's basically Ember Waste anyway so it's cool I guess.

Comment (Hecter94) [reply source/Gamerules.h:37]: > Personally I don't care too much about it except now it's blocking a generally useful PR.
I don't think that's what's blocking this PR. It just needs some more reviews and approvals before merging.

**Comment by Hurleveur** [comment]
Since https://github.com/endless-sky/endless-sky/pull/8046 got merged, this does not change anything with the game (that other PR turned the universal ramscoop back on by default)
I think that the plan is to disable it for the EW which would make sense I think (but for another PR, ofc); anyway asking for @Zitchas opinion again because I believe it should resolve this disagreement we had in the community (at least, that was my hope when opening that PR)

**Comment by Zitchas** [comment]
> Since #8046 got merged, this does not change anything with the game (that other PR turned the universal ramscoop back on by default) I think that the plan is to disable it for the EW which would make sense I think (but for another PR, ofc); anyway asking for @Zitchas opinion again because I believe it should resolve this disagreement we had in the community (at least, that was my hope when opening that PR)

Thanks for working on the mechanics that enabled a compromise to be possible, Hurleveur. It is much appreciated. PR #8201 has been submitted which makes use of your work, with some adjustments to keep things functional. With those mechanics being put to use, my objections here are lifted.

**Review by Zitchas — APPROVED** [review]
Approving the concept of having gamerules, and revoking my prior request for changes as Hurleveur made mechanics to provide us with a compromise situation, and #8201 putting it into action, I'm ok with it. I don't like it, but I'll accept Amazinite's will in this and accept that universal ramscoop is effectively a baked-in part of our lore and that's the way things go.

note: I'm not approving the code, I lack the skill to do that, and I can see there are conflicts to be resolved. I'm approving the concept, purpose, and functionality.

**Comment by Terin** [comment]
Now that we have spaceport checkpoints in as well, that works for me for letting the built-in ramscoop go in the Ember Wastes, as well.

File: source/DataNode.cpp
Comment (Amazinite) [reply source/DataNode.cpp:246]: If by that you mean that adding something like "invisible" to a mission means "invisible = true," then yeah, but we also didn't have functions like these that made having "true" and "false" tokens make sense. And gamerules is a perfect place to have explicit "false" boolean values, as we want the key to be visible so that people know it exists, instead of being forced to just not list the key in order to have a false value.

Comment (1010todd) [reply source/DataNode.cpp:246]: In missions, maybe. It sounds a little odd here though, like `"universal ramscoop" never`, `"death benefits" never`, `"universal ramscoop" always`, `"death benefits" always`. That kinda make it sounds like there's something in between, could it be "sometimes"? or "50%"? While "true" or "false" is just that, either true or false, nothing else in between. Since this is another mechanic to missions entirely we probably don't necessary have to maintain that consistency. Unless we want to have 50% chance you'll get universal ramscoop today or something.

File: data/gamerules.txt
Comment (warp-core) [inline data/gamerules.txt:29]: This should actually probably be `period`, not `rate` at all.

File: source/DataNode.cpp
Comment (Amazinite) [reply source/DataNode.cpp:240]: Because this is copy-pasted from Value.
```suggestion
	return false;
```

**Review by warp-core — APPROVED** [review]

**Pull request merged**
